lib/vrl/parser/src/ast.rs

```diff
@@ -0,0 +1,1132 @@
+use crate::lex::Error;
+use diagnostic::Span;
+use ordered_float::NotNan;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::iter::IntoIterator;
+use std::ops::Deref;
+use std::str::FromStr;
+
+// -----------------------------------------------------------------------------
+// node
+// -----------------------------------------------------------------------------
+
+/// A wrapper type for a node, containing span details of that given node as it
+/// relates to the source input from which the node was generated.
+#[derive(Clone, Eq, Ord, PartialOrd, Hash)]
+pub struct Node<T> {
+    pub(crate) span: Span,
+    pub(crate) node: T,
+}
+
+impl<T> Node<T> {
+    pub fn map<R>(self, mut f: impl FnMut(T) -> R) -> Node<R> {
+        let Node { span, node } = self;
+
+        Node {
+            span,
+            node: f(node),
+        }
+    }
+
+    pub fn new(span: Span, node: T) -> Self {
+        Self { span, node }
+    }
+
+    /// Get a copy of the [`Span`] of the node.
+    pub fn span(&self) -> Span {
+        self.span
+    }
+
+    /// Get the starting byte of the node within the input source.
+    pub fn start(&self) -> usize {
+        self.span.start()
+    }
+
+    /// Get the ending byte of the node within the input source.
+    pub fn end(&self) -> usize {
+        self.span.end()
+    }
+
+    /// Get a reference to the inner node type `T`.
+    pub fn inner(&self) -> &T {
+        &self.node
+    }
+
+    // Consume the node, taking out the [`Span`] and inner node type `T`.
+    pub fn take(self) -> (Span, T) {
+        (self.span, self.node)
+    }
+
+    /// Consume the node, and get the inner node type `T`.
+    pub fn into_inner(self) -> T {
+        self.node
+    }
+
+    /// Consume the node and return a tuple consisting of the start, node type
+    /// `T` and the end position.
+    pub fn into_spanned(self) -> (usize, T, usize) {
+        let Self { span, node } = self;
+
+        (span.start(), node, span.end())
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Node<T> {
```

Comment:
I don't think Debug should hide the Span information.

Reply:
In that case it could be `#[derive]`d.

Reply:
Yeah, I specifically ignored the wrapper because I don't care about it in the case of the AST. It helps simplify testing.

I get how it might confuse others though, so I'll take a look at making it a bit easier to know there's a span there, without using the (too verbose) default derive'd variant.